src/item_factory.cpp

```diff
@@ -869,11 +876,14 @@ typename std::enable_if<std::is_same<T, std::string>::value, bool>::type assign(
 }
 
 template <typename T>
-typename std::enable_if<std::is_same<T, std::set<std::string>>::value, bool>::type assign(
-    JsonObject &jo, const std::string& name, T& val ) {
+typename std::enable_if<std::is_constructible<T, std::string>::value, bool>::type assign(
+    JsonObject &jo, const std::string& name, std::set<T>& val ) {
 
     if( jo.has_string( name ) || jo.has_array( name ) ) {
-        val = jo.get_tags( name );
+        val.clear();
+        for( const auto &e : jo.get_tags( name ) ) {
```

Comment:
You need to add the same loop on line 895 below when inserting from the "extend" object. Currently it tries to `insert( set<string>::iterator, set<string>::iterator )` into a set of `string_id`.

Alternatively use wrapper that converts the set:
```C++
template<typename T>
std::set<T> get_tags( JsonObject &jo, const std::string &name ) {
    std::set<T> result;
    for( auto &t : jo.get_tags( name ) ) {
        result.insert( T( t ) );
    }
    return result;
}

template <typename T>
typename std::enable_if<std::is_constructible<T, std::string>::value, bool>::type assign(
    JsonObject &jo, const std::string& name, std::set<T>& val ) {

    if( jo.has_string( name ) || jo.has_array( name ) ) {
        val = ::get_tags<T>( jo, name );
```

Or make the `get_tags` function in `JsonObject` generic:
```C++
class JsonObject {
        template<typename T = std::string>
        std::set<T> get_tags(const std::string &name);
};
// ... definition of JsonArray comes here, as it's needed by the function. 

template<typename T>
inline std::set<T> JsonObject::get_tags(const std::string &name)
{
    ... // code from the current implementation in the cpp file, slightly adjusted to use T instead of string
}
```
You can call it like `faults = jo.get_flags<string_id<fault>>( "faults" )`